src/api/key_store.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use alloc::vec::Vec;
+use crypto::ecdsa::SecKey;
+use persistent_store::StoreError;
+use rng256::Rng256;
+
+use crate::env::Env;
+
+/// Provides storage for secret keys.
+///
+/// Implementations may use the environment store: [`STORE_KEY`] is reserved for this usage.
+pub trait KeyStore {
+    /// Returns the AES key for key handles encryption.
+    fn kh_encryption(&mut self) -> Result<[u8; 32], Error>;
+
+    /// Returns the key for key handles authentication.
+    fn kh_authentication(&mut self) -> Result<[u8; 32], Error>;
+
+    /// Derives an ECDSA private key from a seed.
+    ///
+    /// The result is big-endian.
+    fn derive_ecdsa(&mut self, seed: &[u8; 32]) -> Result<[u8; 32], Error>;
+
+    /// Generates a seed to derive an ECDSA private key.
+    fn generate_ecdsa_seed(&mut self) -> Result<[u8; 32], Error>;
+
+    /// Resets the key store.
+    fn reset(&mut self) -> Result<(), Error>;
+}
+
+/// Key store errors.
+///
+/// They are deliberately indistinguishable from each other to avoid leaking information.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct Error;
+
+/// Key of the environment store reserved for the key store.
+pub const STORE_KEY: usize = 2046;
+
+impl<T: Env> KeyStore for T {
+    fn kh_encryption(&mut self) -> Result<[u8; 32], Error> {
+        Ok(get_master_keys(self)?.encryption)
+    }
+
+    fn kh_authentication(&mut self) -> Result<[u8; 32], Error> {
+        Ok(get_master_keys(self)?.authentication)
+    }
+
+    fn derive_ecdsa(&mut self, seed: &[u8; 32]) -> Result<[u8; 32], Error> {
+        match SecKey::from_bytes(seed) {
+            None => Err(Error),
+            Some(_) => Ok(*seed),
+        }
+    }
+
+    fn generate_ecdsa_seed(&mut self) -> Result<[u8; 32], Error> {
+        let mut seed = [0; 32];
+        SecKey::gensk(self.rng()).to_bytes(&mut seed);
```

Comment:
Curious about your choice of calling `SecKey::gensk` here and not just storing the seed. The function is not used here yet, but I assume we pay `derive_ecdsa` on assertion and `generate_ecdsa_seed` on make credential, and this way, the assertion is cheaper? Or is it because we don't have the API to generate from seed, and instead we need to pass `self.rng()` in?

Reply:
It's the second. What to do if the seed is not a valid ecdsa key? We could use some seeded DRBG until we find one. But since we have a RNG, we can instead only generate valid seeds. And as you say, this is also more performant (favoring assertion over registration).